---
title: "RNASeq-Proj1"
author: "Daniel Scheuermann"
date: "2024-01-31"
output: html_document
---

```{r initial_setup, include=TRUE}

knitr::opts_chunk$set(echo = TRUE)


# initial setup
library(DESeq2)
library(pheatmap)
library(dplyr)
library(RColorBrewer)
library(ggplot2)
library(ggrepel)
library(clusterProfiler)
library(Annotationdbi)

```

```{r data_setup}

count_data <- read.csv("data/count_matrix.csv", header = TRUE, row.names = 1)

# here you can see the data set directly split into the treated and untreated samples.

# these two separate samples can be compared to show the effects that 'Pasilla' gene depletion has on gene expression.

colnames(count_data)

head(count_data)

# sample information

sample_info <- read.csv("data/design.csv", header = TRUE, row.names = 1)

colnames(sample_info)

head(sample_info)

# the factors here are single vs paired, treated vs untreated.

# setting factor levels
sample_info$Treatment <- factor(sample_info$Treatment)
sample_info$Sequencing <- factor(sample_info$Sequencing)

```

```{r DESeq2_setup}

# creating the deseq object
dds <- DESeqDataSetFromMatrix(countData = count_data, colData = sample_info, design = ~Sequencing + Treatment)

# treatment factor reference
dds$Treatment <- factor(dds$Treatment, level = c("untreated","treated"))

```

```{r data_manip}

# filtering the genes

# we are looking at genes whose count is greater than 10
keep <- rowSums(counts(dds)>10) >= min(table(sample_info$Treatment)) 

dds <- dds[keep,]

# identifying differentially expressed genes
dds <- DESeq(dds)
deseq_results <- results(dds)
deseq_results

# transfer the deseq results into a dataframe
deseq_results <- as.data.frame(deseq_results)
class(deseq_results)

# ordering
deseq_results <- deseq_results[order(deseq_results$pvalue),]

head(deseq_results)
```

```{r data_queries}

# looking specifically at FBgn0039155
deseq_results["FBgn0039155",]

# very low p value and negative log2FoldChange, so this gene is differentially expressed in low levels.

# testing whether or not the Pasilla gene is downregulated by the RNA treatment.
deseq_results["FBgn0261552",]

# also low p value and negative log2FoldChange, so we can conclude this gene is downregulated.

```


```{r data_manip2}

# filtering

# selecting the diferentially expressed genes (alpha = 0.05)
filtered <- deseq_results %>% filter(deseq_results$padj < 0.05)

filtered <- filtered %>% filter(abs(filtered$log2FoldChange) > 1)

head(filtered)

# number of diferentially expressed genes
dim(filtered)

```

```{r queries2}

# save manipulated data in new csv files
write.csv(deseq_results, 'data/deseq_result.all.csv')
write.csv(filtered, 'data/deseq_result.filtered.csv')

normalized_counts <- counts(dds, normalized = TRUE)
head(normalized_counts)
write.csv(normalized_counts, 'data/normalized_counts.csv')
```

```{r visualization}

# dispersion plot
plotDispEsts(dds)

# principal component analysis plot

# variance stabilization
vsd <- vst(dds, blind = FALSE)

# pca plot
plotPCA(vsd, intgroup = c("Sequencing", "Treatment"))

# it seems like there's a clear separation between treated and untreated samples. this demonstrated that there's a path in the data showing a difference in treated vs untreated data samples. The same is true for pairwise vs single samples.

```


```{r heatmaps}

# generate distance matrix
sample_dists <- dist(t(assay(vsd)))
sample_dists_matrix <- as.matrix(sample_dists)
colnames(sample_dists_matrix)

# choose color scheme
colors <- colorRampPalette(rev(brewer.pal(9,"Blues")))(255)

# generated the heat map
pheatmap(sample_dists_matrix, clustering_distance_rows = sample_dists, clustering_distance_cols = sample_dists, color = colors)

# based on the heatmap, it seems like GSM461180_treated_paired sample is similar to GSM461181_treated_pairs sample. Meanwhile, the rest of the samples seem to be relatively different.



# heatmap of top 10 log transformed normalized counts
# sort by the log transformed normalized counts and select first 10 values
top_hits <- deseq_results[order(deseq_results$padj),][1:10,]
top_hits <- row.names(top_hits)
top_hits

# log transformation
rld <- rlog(dds, blind = FALSE)

# heatmap without clustering
pheatmap(assay(rld)[top_hits,], cluster_rows = FALSE, show_rownames = TRUE, cluster_cols = FALSE)

# heatmap with clustering
pheatmap(assay(rld)[top_hits,])

# adding annotations
annot_info <- as.data.frame(colData(dds)[,c("Sequencing","Treatment")])

# heatmap with annotations
pheatmap(assay(rld)[top_hits,], cluster_rows = FALSE, show_rownames = TRUE, cluster_cols = FALSE, annotation_col = annot_info)



# heatmap of z-scores

# calculating z-scores
cal_z_score <- function(x) {(x-mean(x)) / sd(x)}

z_scores <- t(apply(normalized_counts, 1, cal_z_score))
z_score_subset <- z_scores[top_hits,]
pheatmap(z_score_subset)



# MA plot

plotMA(dds, ylim = c(-2,2))

# removing noise
resLFC <- lfcShrink(dds, coef = "Treatment_treated_vs_untreated", type = "apeglm")
plotMA(resLFC, ylim = c(-2,2))



# Volcano plot
# changed the resLFC to a data frame to make a plot with little noise
resLFC <- as.data.frame(resLFC)

# labeling genes (diferentially expressed vs otherwise)
resLFC$diffexpressed <- "NO"
resLFC$diffexpressed[resLFC$log2FoldChange > 0.1 & resLFC$padj < 0.05] <- "UP"
resLFC$diffexpressed[resLFC$log2FoldChange < 0.1 & resLFC$padj < 0.05] <- "DOWN"

resLFC$delabel <- NA

# volcano plot (note noise is removed)
ggplot(data = resLFC, aes(x = log2FoldChange, y = log10(pvalue), col = diffexpressed, label = delabel)) +
  geom_point() +
  theme_minimal() +
  geom_text_repel() +
  scale_color_manual(values = c("blue","black","red")) +
  theme(text = element_text(size = 20))
```

```{r gene_ontology}


